app-modules/meeting-center/src/Managers/OutlookCalendarManager.php

```diff
@@ -31,65 +31,281 @@
+        } catch (ClientException $exception) {
+            if ($exception->getCode() === 401) {
+                $calendar = $this->refreshToken($event->calendar);
 
-        // TODO: Implement updateEvent() method.
+                $request->setAccessToken($calendar->oauth_token);
+
+                $response = $request->execute();
+            } else {
+                throw $exception;
+            }
+        }
+
+        $event->provider_id = $response->getResponseAsObject(Event::class)->getId();
+        $event->saveQuietly();
     }
 
     public function deleteEvent(CalendarEvent $event): void
     {
-        // https://learn.microsoft.com/en-us/graph/api/event-delete?view=graph-rest-1.0&tabs=http
+        $client = $this->makeClient($event->calendar);
+
+        $request = $client->createRequest(
+            requestType: 'DELETE',
+            endpoint: "/me/calendars/{$event->calendar->provider_id}/events/{$event->provider_id}",
+        );
+
+        try {
+            $request->execute();
+        } catch (ClientException $exception) {
+            if ($exception->getCode() === 401) {
+                $calendar = $this->refreshToken($event->calendar);
+
+                $request->setAccessToken($calendar->oauth_token);
 
-        // TODO: Implement deleteEvent() method.
+                $request->execute();
+            } else {
+                throw $exception;
+            }
+        }
     }
 
     public function syncEvents(Calendar $calendar, ?Datetime $start = null, ?Datetime $end = null, ?int $perPage = null): void
     {
-        // TODO: Implement syncEvents() method.
+        $providerEvents = collect($this->getEvents($calendar, $start, $end, $perPage));
+
+        $providerEvents
+            ->each(function (Event $providerEvent) use ($calendar) {
+                $userEvent = $calendar->events()->where('provider_id', $providerEvent->getId())->first() ?? $calendar->events()->make();
+
+                $userEvent->fill([
+                    'provider_id' => $providerEvent->getId(),
+                    'title' => $providerEvent->getSubject(),
+                    'description' => $providerEvent->getBodyPreview(),
+                    'starts_at' => Carbon::parse($providerEvent->getStart()->getDateTime(), $providerEvent->getStart()->getTimeZone()),
+                    'ends_at' => Carbon::parse($providerEvent->getEnd()->getDateTime(), $providerEvent->getEnd()->getTimeZone()),
+                    'attendees' => collect($providerEvent->getAttendees())
+                        ->map(fn ($attendee) => $attendee['emailAddress']['address'])
+                        ->prepend($calendar->provider_email),
+                ]);
+
+                if ($userEvent->isDirty()) {
+                    $userEvent->saveQuietly();
+                }
+            });
+
+        $calendar->events()
+            ->whereNull('provider_id')
+            ->each(fn ($event) => $this->createEvent($event));
+
+        $calendar->events()
```

Comment:
I can't quite tell, will this delete past events like the problem we had with Google?

Reply:
It will not since we return all events within the time span we set. So it will get the events in the past as well here and not delete them.